### ISSUE TITLE
torch.utils.checkpoint is not compatible with nn.DataParallel

### ISSUE THREAD

**Issue by Lyken17** [pytorch/pytorch#13234]
## To Reproduce

```
net = nn.Sequential(
	# somenet
)
from torch.utils.checkpoint import checkpoint_sequential

net = nn.DataParallel(net)
data = torch.randn(1, 3, 224, 224, requires_grad=True)
out = checkpoint_sequential(net, 1, data) # error here
```

## Expected behavior
Checkpoint sequential is expected to work well with nn.DataParallel. 

## Environment
pytorch: 0.4
OS: ubuntu 16.04

**Comment by davidmascharka** [comment]
Calling `DataParallel` wraps the model:

``` python
import torch

net = torch.nn.Sequential(
    torch.nn.Conv2d(3, 1, 3)
)
print(net)

Sequential(
  (0): Conv2d(3, 1, kernel_size=(3, 3), stride=(1, 1))
)
```

When you wrap it in `DataParallel`, it becomes:

``` python
net = torch.nn.Sequential(
    torch.nn.Conv2d(3, 1, 3)
)

net = torch.nn.DataParallel(net)
print(net)

DataParallel(
  (module): Sequential(
    (0): Conv2d(3, 1, kernel_size=(3, 3), stride=(1, 1))
  )
)
```

When you make your `checkpoint_sequential` call, you aren't passing in a `Sequential` model; you're passing in a `DataParallel` model. You can call `checkpoint_sequential(net.module, 1, data)` to get that functionality.
```

**Comment by Lyken17** [comment]
Hi @davidmascharka , thanks for your reply but I think it does not utilize multi gpus.

**Comment by davidmascharka** [comment]
Correct. You either need to wrap the model in `DataParallel` after loading or you need to use `checkpoint` instead of `checkpoint_sequential`.

**Comment by mkelcb** [comment]
> Correct. You either need to wrap the model in `DataParallel` after loading or you need to use `checkpoint` instead of `checkpoint_sequential`.
@davidmascharka  What do you mean by 'after loading'? What is the correct schema to apply DataParallel and checkpoint_sequential together so that it works on multiple GPUs?
